### PR TITLE
Certora verification of the DLL

### PR DESCRIPTION
**Pull request by QGarchery**

### CODE REVIEW

File: certora/specs/dll-simple.spec [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    }
+
+// Derived results
+
+// result: isForwardLinkedBetween(getHead(), getTail())
+// explanation: if getTail() == 0, then from tipsZero() we know that getHead() == 0 so the result follows
+// otherwise, from tailWellFormed(), we know that inDLL(getTail()) so the result follows from DLLisForwardLinked(getTail()).
+
+// result: forall addr. isForwardLinkedBetween(addr, getTail())
+// explanation: it can be obtained from the previous result and DLLisForwardLinked.
```
Comment (MathisGD) [inline certora/specs/dll-simple.spec:281]: Intuitively it is ok, but I'm not sure how would you prove that

File: certora/specs/dll-simple.spec [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+// result: "BackwardLinked" dual results
+// explanation: it can be obtained from ForwardLinked and twoWayLinked.
+
+// result: there is only one list
+// explanation: it comes from the fact that every non zero address is linked to getHead().
```
Comment (MathisGD) [inline certora/specs/dll-simple.spec:287]:
```suggestion
// explanation: it comes from the fact that every non zero address that is in the DLL is linked to getHead().
```

**Review by MathisGD — COMMENTED** [review]
Amazing ! I mostly have questions, can we schedule a call so that you can explain me everything ?

File: certora/applyHarnessSimple.patch [hunk 3]
```diff
@@ -0,0 +1,251 @@
+diff -ruN DLL.sol DLL.sol
+--- DLL.sol	1970-01-01 01:00:00.000000000 +0100
++++ DLL.sol	2022-09-30 14:57:52.658495991 +0200
+@@ -0,0 +1,87 @@
++// SPDX-License-Identifier: GNU AGPLv3
++pragma solidity ^0.8.0;
++
++import "./DoubleLinkedList.sol";
++
++contract DLL {
```
Comment (MathisGD) [inline certora/applyHarnessSimple.patch:10]: It could be great to harmonize name (concrete..)

File: certora/scripts/sanity-simple.sh [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
+
+make -C certora munged
+
+certoraRun \
+    certora/munged/DLL.sol \
+    --verify DLL:certora/specs/sanity.spec \
+    --loop_iter 3 \
```
Comment (MathisGD) [inline certora/scripts/sanity-simple.sh:8]:
```suggestion
    --loop_iter 7 \
```

File: certora/specs/dll-simple.spec [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+    isHeadWellFormed()
+    { preserved remove(address rem) {
+        requireInvariant zeroEmpty();
+        requireInvariant twoWayLinked(getPrev(rem), rem);
+        requireInvariant twoWayLinked(rem, getNext(rem));
+        requireInvariant linkedIsInDLL(getNext(rem));
+      }
+    }
+
+invariant tailWellFormed()
```
Comment (MathisGD) [inline certora/specs/dll-simple.spec:68]: Maybe add a comment

File: certora/applyHarnessSimple.patch [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+diff -ruN test/TestRedBlackBinaryTree.sol test/TestRedBlackBinaryTree.sol
+--- test/TestRedBlackBinaryTree.sol	2022-05-31 14:28:35.619981325 +0200
++++ test/TestRedBlackBinaryTree.sol	1970-01-01 01:00:00.000000000 +0100
+@@ -1,50 +0,0 @@
+-// SPDX-License-Identifier: UNLICENSED
+-pragma solidity ^0.8.0;
+-
+-import "../RedBlackBinaryTree.sol";
+-
+-contract TestRedBlackBinaryTree {
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply certora/applyHarnessSimple.patch:207]: Resolving this as it should be done along with #54

File: certora/specs/dll-simple.spec (hunk 5 above)
Comment (QGarchery) [reply certora/specs/dll-simple.spec:68]: I don't remember this, a comment about what ?

Comment (MathisGD) [reply certora/specs/dll-simple.spec:68]: about what it is verifying: construction and remove

**Comment by Rubilmax** [comment]
Is this PR ready to be reviewed?

**Comment by QGarchery** [comment]
See #97 instead

**Pull request closed without merging**